R/tribble.R

```diff
@@ -27,19 +27,75 @@
 #'   "b", 4:6
 #' )
 tribble <- function(...) {
+  data <- extract_frame_data_from_dots(...)
 
+  if (length(data$frame_rest) == 0) {
+    out <- rep(list(logical()), length(data$frame_names))
+    names(out) <- data$frame_names
+    return(as_tibble(out))
+  }
+
+  frame_mat <- matrix(data$frame_rest, ncol = length(data$frame_names), byrow = TRUE)
+
+  frame_col <- lapply(seq_len(ncol(frame_mat)), function(i) {
+    col <- frame_mat[, i]
+    if (any(vapply(col, needs_list_col, logical(1L)))) {
+      col
+    } else {
+      unlist(col)
+    }
+  })
+
+  # Create a tbl_df and return it
+  names(frame_col) <- data$frame_names
+  as_tibble(frame_col)
+}
+
+frame_matrix <- function(...) {
+  data <- extract_frame_data_from_dots(...)
+
+  if (any(vapply(data$frame_rest, needs_list_col, logical(1)))) {
```

Comment:
Same here -- please extract everything after `extract_frame_data_from_dots(...)` to a function.